src/utils.jl

```diff
@@ -1,5 +1,24 @@
 using Base: ImmutableDict
 
+function lockbuffer!(f, params)
+  if params.threaded
+    lock(params.buffer_lock) do
+      return f()
+    end
+  else
+    return f()
+  end
+end
+function lockmergesbuffer!(f, params)
+  if params.threaded
+    lock(params.merges_buffer_lock) do
+      return f()
+    end
+  else
+    return f()
+  end
+end
+
```

Comment:
Why lock the buffer only if threaded?

Reply:
Do we need to lock if not threaded? The docs are fairly clear that you shouldn't take locks in generated functions (although this is *likely* to be problematic mostly for globals I was told it's probably best to avoid them even when passed in).

Reply:
I switched the name of this parameter in the EGraph to `needslock`. We *don't* need to lock within a generated function, so I'd like to keep that functionality. But that separates it out from threading.

It's also not clear to me what `params.threaded` does? It doesn't seem to be used anywhere.